### ISSUE TITLE
ucm should gracefully handle a locked sqlite db

### ISSUE THREAD

**Issue by stew** [unisonweb/unison#2622]
for our current experiments with unison cloud, we are finding ourselves running multiple ucm processes against the same codebase at the same time. For the most part this works, but sometimes it seems that two of them try to lock the DB at the same time, which results in the loser crashing with an error:

    Exception: SQLite3 returned ErrorBusy while attempting to perform step: database is locked

It would be great if we could recover from these errors, since restarting ucm takes a long time with our current codebase

**Comment by aryairani** [comment]
@stew Could you elaborate on what's going on in the two ucms when this happens?

**Comment by ChrisPenner** [comment]
Yeah we definitely need to take a pass over our DB layer and improve the transactionality of  everything; we've been talking about it, but if you're running into this more often maybe the time has come.

It'd probably also be nice if there was a mechanism for transactional operations at the level of the Codebase abstraction; not sure if we need it just yet, but any time you're executing a write/delete on the codebase as a result of some codebase read we should be doing that transactionally.

**Comment by mitchellwrosen** [comment]
Yeah, I agree with Chris.

I think the first step here is to migrate our queries over to using a `Transaction` monad that has some reasonable retry-on-busy logic. That would at least cover the bug report here. I've started on this.

Next step would be to make running multiple ucm processes simultaneously actually work, which would involve consulting the `data_version` wherever necessary, to notice when the database has been modified by another connection (i.e. process).

**Comment by stew** [comment]
@aryairani if you have two UCMs that are watching the same directory and the same codebase and create a watch expression for example, often one of the two will crash

**Comment by ChrisPenner** [comment]
We also just identified that we're often not closing database connections correctly;
I've got a PR open for some in the lower level here, #2641 ; but I think we might still not actually be closing DB connections properly when users `Ctrl-C` for example 😬 .

Regardless, we still should have some way to roll back and try again on conflicts.

**Comment by ChrisPenner** [comment]
My initial thoughts here would be:
* Add some mechanism/exception/handler for detecting and catching this specific error
* Wrap every iteration of the command-loop in its own delayed transaction
* Catch contention errors in the command-loop, if it’s detected, roll-back the command’s transaction and just run the command again (maybe after a small backoff)
* I’ll need to go through and ensure all the IO in each command is idempotent or okay to be duplicated, As part of this I’ll likely remove MonadIO in a separate PR and add a transactionalIO :: IO a -> m a  lifter with docs specifying all the caveats of doing IO in a transaction.

**Comment by aryairani** [comment]
blocked on #2718
